### PR TITLE
zero size allocations

### PR DESCRIPTION
**Pull request by BenjaminW3**
This removes the assertions that did not allow to create zero size CPU and CUDA buffers and to delete a nullptr via `cudaFree` (issue #182).
Furthermore, a test verifying that the basic view operations are valid on such a (multidimensional) zero size buffer has been added.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
Please try to pin a zero alloced host buffer! This should be fail because cuda not allow to pin zero byte memory.
A solution is to check the size while the memory should be  pinned.

**Comment by BenjaminW3** [comment]
I have added an additional check.

**Comment by BenjaminW3** [comment]
I will resolve the merge conflicts this evening.

File: include/alpaka/mem/buf/BufCpu.hpp [hunk 1]
```diff
@@ -519,6 +518,11 @@ namespace alpaka
                         if(!mem::buf::isPinned(buf))
                         {
 #if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && defined(__CUDACC__)
+                            if(buf.m_extentElements.prod() == 0u)
+                            {
+                                throw std::runtime_error("Zero size buffers can not be pinned!");
```
Comment (psychocoderHPC) [inline include/alpaka/mem/buf/BufCpu.hpp]: Is this necessary to throw an error? We can ignore the pinning call if the memory contains zero elements.
The reason for that is that the memory behind a address is pinned. But there is no memory (zero elements) therefore not pinning is a clean and valid solution.

Comment (BenjaminW3) [inline include/alpaka/mem/buf/BufCpu.hpp]: You are right. I will remove the throw.

**Comment by BenjaminW3** [comment]
The AppVeyor build hit the time limit...

**Comment by psychocoderHPC** [comment]
Big Thanks for this nice pull request!

**Pull request merged**